src/mlir/dialect/VeronaTypes.h

```diff
@@ -64,41 +71,42 @@ namespace mlir::verona
     llvm::ArrayRef<mlir::Type> getElements() const;
   };
 
-  struct IntegerType
-  : public Type::TypeBase<IntegerType, Type, detail::IntegerTypeStorage>
-  {
-    using Base::Base;
-
-    static IntegerType get(MLIRContext* context, size_t width, unsigned sign);
-
-    size_t getWidth() const;
-    bool getSign() const;
-  };
-
-  struct FloatType
-  : public Type::TypeBase<FloatType, Type, detail::FloatTypeStorage>
-  {
-    using Base::Base;
-
-    static FloatType get(MLIRContext* context, size_t width);
-
-    size_t getWidth() const;
-  };
-
-  struct BoolType : public Type::TypeBase<BoolType, Type, TypeStorage>
+  /**
+   * Unknown types are derived types from operations that cannot define the type
+   * at lowering stage, but will later be replaced by other types during type
+   * inference. Not all types will be concrete by then, but none of them should
+   * be unknown.
+   */
+  struct UnknownType : public Type::TypeBase<UnknownType, Type, TypeStorage>
   {
     using Base::Base;
 
-    static BoolType get(MLIRContext* context);
+    static UnknownType get(MLIRContext* context);
   };
 
-  struct UnknownType : public Type::TypeBase<UnknownType, Type, TypeStorage>
+  /**
+   * A class descriptor type, used for access to static members of the class,
+   * including fields and methods. The type pointed to could be unknown before
+   * type inference, but once known, should point to a fully qualified class.
+   */
+  struct DescriptorType
+  : public Type::TypeBase<DescriptorType, Type, detail::DescriptorTypeStorage>
   {
     using Base::Base;
 
-    static UnknownType get(MLIRContext* context);
+    static DescriptorType get(MLIRContext* context, Type ref);
+    TypeRange getTypes() const;
   };
 
+  /**
+   * Capability types represents traits that other types possess.
+   *
+   * Isolated: An entry point to a new region, can only have one reference to
+   * it at any time (ownership model). Can also be created when returning a
+   * sub-region that needs to be isolated.
+   *
+   * Mutable/Immutable: Read-Write/Read-Only variables and viewpoints.
```

Comment:
`readonly` is distinct from immutable.  An `imm` capability refers to an immutable object, a `readonly` capability refers to an object that cannot be modified via this pointer (similar to C/C++ `const`, though stricter) but may or may not actually be immutable.

Reply:
That's what I meant as the difference between "variables and viewpoints". IIUC, an `imm` variable is always immutable, while a variable seen from an `imm` viewpoint is only read-only through that "pointer".

Perhaps I should be clearer...

Reply:
I hope by being less verbose it got clearer...